### PR TITLE
Frogger task

### PR DESCRIPTION
**Pull request by Tedzury**
### Frogger game

[Demo](https://tedzury.github.io/FroggerGame/) |
[Code base](https://github.com/Tedzury/FroggerGame)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by Tedzury** [comment]
Good evening! Applied minor fixes after self check. Ready to ask for review :)

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
+const Enemy = function (y, speed) {
+  this.x = -80;
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:2]: What is -80?

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  },
+];
+
+const Player = function (x, y) {
+  this.x = x;
+  this.y = y;
+  this.moveX = 100;
+  this.moveY = 90;
+  this.width = 80;
+  this.height = 60;
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:58]: Duplication of code. Make code DRY.

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+Player.prototype.render = function () {
+  ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+};
+
+Player.prototype.handleInput = function () {};
+
+Player.prototype.respawn = function () {
+  this.x = 200;
+  this.y = 400;
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:88]: What are 200 and 400?
The same question to other mystery numbers across the code.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@Tedzury nice try.

As a peer developer I want to understand what is behind numbers. That is why we use named constants and calculate more complex things from basic and comprehensible ones.

As a peer developer I want to make changes to as few lines of code as possible when needed. That is why we write DRY code.

**Comment by Tedzury** [comment]
Good morning. Made slight refactoring, renamed some variables to better readability, tried to remove repeating for width and height, but failed. Assigned an entity as proto to player and it works, but I can't assing the entity as prototype for enemy, it break the app completely, function update stops working. I think it's because we are putting the enemies in the array and somehow it breaks the main update function.

**Comment by OleksiyRudenko** [comment]
> Made slight refactoring, renamed some variables to better readability, 
👍 

> tried to remove repeating for width and height, but failed.
👍 

> Assigned an entity as proto to player and it works, but I can't assing the entity as prototype for enemy, it break the app completely, function update stops working. I think it's because we are putting the enemies in the array and somehow it breaks the main update function.

If done properly it should work.
If you try to extend plain object then it will definitely fail. However it is just my hypotheses. Better told on the code.
Consider two options:
(1) Submit the "wrong" code for `Entity` and we will see where it goes
(2) Seek help in Students' chat

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  this.width = width;
+  this.height = height;
+  this.sprite = "images/enemy-bug.png";
+};
+
+Enemy.prototype.update = function (dt) {
+  this.xAxisPosition += this.moveSpeed * dt;
+
+  if (this.xAxisPosition >= ctx.canvas.width) {
+    this.xAxisPosition = -80;
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:19]: Still a mistery number? Is it related to `entity.width` somehow?

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    yAxisPosition: 60,
+    moveSpeed: 180,
+  },
+  {
+    yAxisPosition: 140,
+    moveSpeed: 140,
+  },
+  {
+    yAxisPosition: 220,
+    moveSpeed: 100,
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:57]: As a peer developer I am tasked to add 5 more enemies.
I want to do it with the least effort possible. The numbers here look mysteriously.
What I see and know is that the play field has clear rows. Ideally, I'd like to define a row for an enemy and let the program calculate pixels.

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+  },
+];
+
+const Player = function (xAxisPosition, yAxisPosition) {
+  this.xAxisPosition = xAxisPosition;
+  this.yAxisPosition = yAxisPosition;
+  this.moveXAxis = 100;
+  this.moveYAxis = 90;
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:65]: Mysterious numbers.

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  this.score = 0;
+  this.scoreMax = 0;
+};
+
+Player.prototype = entity;
+
+Player.prototype.update = function () {
+  if (player.yAxisPosition < 0) {
+    player.respawn();
+    player.win();
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:76]: What is the way to address object instance properties and methods?
This is not a rhetoric question. You may also want to review the module's theoretical materials.

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    this.yAxisPosition
+  );
+};
+
+Enemy.prototype.checkCollision = function () {
+  if (
+    player.xAxisPosition < this.xAxisPosition + this.width &&
+    player.width + player.xAxisPosition > this.xAxisPosition &&
+    player.yAxisPosition < this.yAxisPosition + this.height &&
+    player.yAxisPosition + player.height > this.yAxisPosition
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:38]: From the checklist:
> * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
>      (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
>      Hint: pass `Player` instance as an argument to every enemy

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@Tedzury good progress!
Although code needs further improvements.

**Comment by Tedzury** [comment]
Good evening. I asked for help in students chat and found response :) Totally refactored almost everything. Hope, it is what was  needed, if not - sorry for my dumbness :(

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+const Player = function (xAxisPosition, yAxisPosition, sprite, width, height) {
+  Entity.call(this, xAxisPosition, yAxisPosition, sprite, width, height)
+  this.moveXAxis = tileWidth;
+  this.moveYAxis = tileHeight;
+};
+
+Player.prototype = Object.create(Entity.prototype);
+
+Player.prototype.update = function () {
+  if (player.yAxisPosition < 0) {
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:56]: Why do you refer to a particular `Player` instance here?
Looks weird when you obviously know the proper way of referring to object's properties.

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+Enemy.prototype.checkCollision = function () {
+
+  if (
+    player.xAxisPosition < this.xAxisPosition + this.width &&
+    player.width + player.xAxisPosition > this.xAxisPosition &&
+    player.yAxisPosition < this.yAxisPosition + this.height &&
+    player.yAxisPosition + player.height > this.yAxisPosition
+  ) {
+    
+    player.respawn();
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:135]: > classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
>  (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
>  Hint: pass `Player` instance as an argument to every enemy

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+  {
+    yAxisPosition: firstEnemyRow,
+    moveSpeed: 180,
+  },
+  {
+    yAxisPosition: secondEnemyRow,
+    moveSpeed: 150,
+  },
+  {
+    yAxisPosition: thirdEnemyRow,
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:151]: `firstAnythin`, `secondAnything` etc just like `anything1`, `anything2` clearly calls for an array.
`whatever[1]` or `whatever[natural number - 1]` looks more comprehensible.

Also using arrays will enable `Array` mehods. Imagine you are tasked to add more rows for enemies.

File: submissions/Oleksii_Drohachov/Frogger_task/app.js [hunk 12]
```diff
@@ -0,0 +1,182 @@
+
+const tileWidth = 100;
+const tileHeight = 83;
+const numberOfRows = 6;
+const numberOfColumns = 5;
+const fieldWidth = numberOfColumns * tileWidth;
+const fieldHeight = numberOfRows * tileHeight;
+const fieldBottomGap = 100;
+const emenyRowsModifier = 0.66;
```
Comment (OleksiyRudenko) [inline submissions/Oleksii_Drohachov/Frogger_task/app.js:9]: 0.66 looks weird. Do you research as to sprite images characteristics. Do not rely on some else's calculations/numbers, they may be wrong.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@Tedzury very good job and great improvement since v1!
Yet a few improvements are required.

**Comment by Tedzury** [comment]
Good day. Resolved all mistakes, slightly reafactored player and enemy classes to remove interdependence between them. Now passing a player to enemy as argument, and scoreBoards as argument to player. 

And, what took a lot of of energy - refactored allEmenies, now it's result of function, which returns an array of Enemies.

About your comment "0.66 looks weird. Do your own research of sprite images characteristics. Do not rely on some else's calculations/numbers, they may be wrong." - changed this variable to fixed 19 value. Don't know how to get rid of it, because it is just a visual variable, made to center bugs in center of the track, without it they visually fell down from track they are running.  I weren't cheating, found this number only by own visual inspection.

**Review by OleksiyRudenko — APPROVED** [review]
@Tedzury great job done!

**Pull request merged**